pxtlib/emitter/decompiler.ts

```diff
@@ -304,9 +304,29 @@ ${output}</xml>`;
         }
 
         function emitVariableSetOrChange(name: string, value: ts.Expression | number, changed = false) {
+            const isNumber = typeof value === 'number';
+            if (!isNumber) {
+                const valueNode = value as ts.Expression;
+
+                // Assignments to null are not supported because there is no block value equivalent to null.
+                // However, declarations that initialize to null are fine so long as they occur in the global
+                // scope since blocks that have object types will regenerate the initializations in compilation
+                if (valueNode.kind === SyntaxKind.NullKeyword) {
+                    if (valueNode.parent.kind !== SyntaxKind.VariableDeclaration) {
+                        error(valueNode, "Assigning null to a variable is not supported in blocks");
+                    }
+                    else if (getEnclosingBlockScopeContainer(valueNode).kind !== SyntaxKind.SourceFile) {
+                        error(valueNode, "Only globally scoped initializations to null are supported in blocks");
```

Comment:
Ok sounds good.

As far as decompilation, we should not rely on the rountrip forever. Instead, we should detect unsupported features and bail out. The rountrip was mostly there to catch a lot of the issues. It brings its own set of challenges (whitespace, etc...) as well. But that's beyond the point of this PR.